### DISCUSSION TITLE
Integration with Jumpstart Rails—how possible is it?

### DISCUSSION THREAD

**Discussion by brandonzylstra** [avo-hq/avo#992] [Help] [answered]
I'm looking into building an app with Jumpstart Rails, but I'd like to use Avo instead of Administrate. (JumpStart Rails includes Administrate, but I find Avo far more appealing.)

I realize there is some overlap between what JSR and Avo provide—like authentication.  So I have several questions: 

1. Is Avo flexible enough to let me pick and choose which aspects of it I use? e.g. can I avoid using Pundit without huge pain being inflicted?  ([The word "if" in the section on file authorization](https://docs.avohq.io/2.0/fields.html#authorization) makes me think the answer is "yes" regarding Pundit, but that's just one example, so the first question still holds.)
2. Can I easily skip [both webpacker and sprockets](https://docs.avohq.io/2.0/custom-asset-pipeline.html) and go with Rails 7 defaults?  Or does "or a different one" only include [esbuild](https://esbuild.github.io/) or [rollup.js](https://rollupjs.org/) with [jsbundling-rails](https://github.com/rails/jsbundling-rails)?
3. Can anyone reading this foresee any conflicts that I might not have noticed by reading the docs?
4. Has anyone here tried using them both?

Fortunately both use Tailwind CSS, so the Bootstrap bullet will have been dodged 😅.

**Comment by adrianthedev** [comment, marked as answer]
Hey @brandondrew, thanks for asking all these thorough questions.

### TL;DR;
They work amazing together!

I started work on bringing Avo in Jumpstart as an alternative to administrate. I spoke with Chris already.

1. In general I tried to make it as unopinionated as possible. Of course, as with any pre-packaged piece of software, some things might be tied together and be the defaults (just like you get with Jumpstart and others). Avo needs very few packages (see in the [`gemspec`](https://github.com/avo-hq/avo/blob/main/avo.gemspec)), so most of what you'll build, and how you build it is really up to you. So you can use your own authorization (devise, warden, omnniauth, anything else), your own search engine (we recommend ransack, but you can bring your own), your own templating engine (we stick with the default `erb`, but you go crazy), and so on. Regarding Pundit, that's our default method of authorization (I'll change the wording on the docs) and it's built-in Avo all throughout. If you use CanCanCan or something else, I bet you could create an adapter system that pundit policies, call o CanCanCan config somehow.

2. If you don't want to add any JS/CSS assets to Avo, you're all good. You can start an app with Rails 7 defaults. 
However, if you, want to bring in any JS/CSS, you need an asset pipeline. I believe pretty much any pipeline works. It works beautifully with webpacker and sprockets (haven't tested others but I'm confident they'll work fine). Avo being a Rails engine, it will behave as any other piece of Rails code when it comes to assets.
Avo's assets are abstracted away from our build step. When we package Avo, we build the CSS and JS files (using esbuild because it's fast) and send them bundled with the package so you never have to worry about anything of Avo's, and ship your app without NodeJS (or other build tools Avo would need) as a dependency 🙌
Next, when you need to add your own assets, Rails will just use your pipeline.
There's a guide [here](https://docs.avohq.io/2.0/custom-asset-pipeline.html) and I'm open to answering any questions you might have going forward.

3. Avo is designed to work with pretty much any type of Rails app. 
Of course, there are "fair-usage" boundaries. One example is you can't use it in an `api_only` Rails app because Rails takes out some parts of the framework around views and sessions. But if you're building a "regular" "Rails-way" app, you should be good.
I've been building apps on Avo for about a year now. A lot of features that I bring to Avo are from my experiences and some are from customer feedback, so most of the sensible customer feedback gets into the framework.

4. I am currently using Avo with Jumpstart Rails and they go together just perfect! I even configured the resources to show custom mixed data between subscriptions (from Jumpstart) and licenses (from Avo).


Happy to answer more of your questions✌️

**Reply by brandonzylstra** [reply]
Wow this is amazing news.  Thanks for the detailed answers! 

I would love it if Chris made Avo the default, although as long as it's easy for me to switch to Avo I'm happy.  

Any tips that you can publish about details of how you did it would be appreciated.

**Reply by adrianthedev** [reply]
Yeah, it's going to be easy to switch to Avo in Jumpstart. Probably through the config wizard.

I'll try and write an article about it. Might help the greater Jumpstart community 👍
